File: python/runners/image-to-text/image_to_text.py
"""Describe an image, or answer a question about it, with a locally pulled Hugging Face transformers model.

Compiled into a standalone binary with PyInstaller, so it runs without a Python
installation. Designed to be driven by the PHP wrapper:

  * stdout: one JSON object on success, e.g. {"text": "two cats sleeping on a pink couch"}
  * stderr: progress and error messages for humans, and the text as it is written
  * exit code: 0 success, 1 generation failed, 2 invalid arguments

Both kinds of image-to-text models are supported:
  * captioning models (task "image-to-text"), e.g. BLIP, GIT, ViT-GPT2; a prompt, when given,
    is the start of the caption
  * vision-language models (task "image-text-to-text"), e.g. SmolVLM, Qwen2-VL, LLaVA; they answer
    the prompt about the image through their chat template

Input is handed to transformers as-is: values the model cannot handle make the run fail.
"""

import argparse
import json
import multiprocessing
import os
import sys
import traceback

EXIT_OK = 0
EXIT_FAILED = 1

DEFAULT_MAX_NEW_TOKENS = 256

# What vision-language models are asked when no prompt is given.
DEFAULT_QUESTION = "Describe this image."


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="image-to-text",
        description="Describe an image, or answer a question about it, with a locally pulled transformers model.",
    )
    parser.add_argument("--model", required=True, help="Directory of a pulled transformers model (contains config.json)")
    parser.add_argument("--image", required=True, help="Path of the image file, e.g. photo.jpg")
    parser.add_argument(
        "--prompt",
        help=f"Question about the image for vision-language models (default: '{DEFAULT_QUESTION}'), "
        "or the start of the caption for captioning models",
    )
    parser.add_argument(
        "--max-new-tokens",
        type=int,
        default=DEFAULT_MAX_NEW_TOKENS,
        help=f"Maximum number of tokens to generate (default: {DEFAULT_MAX_NEW_TOKENS})",
    )
    parser.add_argument("--temperature", type=float, help="Sampling temperature; 0 always picks the likeliest token (default: the model's own)")
    parser.add_argument("--seed", type=int, help="Random seed, for reproducible text")
    parser.add_argument("--device", help="Torch device, e.g. cpu, cuda, mps (default: the best available)")
    return parser.parse_args(argv)


def log(message: str) -> None:
    print(message, file=sys.stderr, flush=True)


def main(argv: list[str]) -> int:
    args = parse_args(argv)

    # Models are loaded from disk only; never reach out to the Hub.
    os.environ["HF_HUB_OFFLINE"] = "1"

    # Imported after argument parsing so --help and usage errors stay fast.
    import torch
    from PIL import Image
    from transformers import (
        AutoImageProcessor,
        AutoModelForImageTextToText,
        AutoProcessor,
        AutoTokenizer,
        PreTrainedTokenizerBase,
        TextStreamer,
    )
    from transformers.utils import logging as transformers_logging

    # Progress bars redraw with carriage returns, which garbles pipes and log files.
    if not sys.stderr.isatty():
        transformers_logging.disable_progress_bar()

    class StderrStreamer(TextStreamer):
        """Writes the text to stderr as it is generated, keeping stdout for the result."""

        def on_finalized_text(self, text: str, stream_end: bool = False) -> None:
            print(text, file=sys.stderr, flush=True, end="" if not stream_end else "\n")

    try:
        device = args.device or best_device(torch)
        dtype = torch.float16 if device.startswith("cuda") else torch.float32

        image = Image.open(args.image).convert("RGB")

        log(f"[image-to-text] Loading {args.model} on {device}...")
        processor = AutoProcessor.from_pretrained(args.model, local_files_only=True)
        if isinstance(processor, PreTrainedTokenizerBase):
            # Models without a combined processor (e.g. ViT-GPT2) ship an image processor and a tokenizer separately.
            tokenizer, processor = processor, AutoImageProcessor.from_pretrained(args.model, local_files_only=True)
        else:
            tokenizer = getattr(processor, "tokenizer", None) or AutoTokenizer.from_pretrained(args.model, local_files_only=True)

        model = AutoModelForImageTextToText.from_pretrained(args.model, dtype=dtype, local_files_only=True).to(device)
        model.eval()

        chat = bool(getattr(processor, "chat_template", None))
        if chat:
            messages = [{"role": "user", "content": [{"type": "image"}, {"type": "text", "text": args.prompt or DEFAULT_QUESTION}]}]
            text = processor.apply_chat_template(messages, add_generation_prompt=True)
            inputs = processor(text=text, images=[image], return_tensors="pt")
        elif args.prompt and hasattr(processor, "tokenizer"):
            # The prompt is the start of the caption, e.g. "a photography of".
            inputs = processor(images=image, text=args.prompt, return_tensors="pt")
        else:
            inputs = processor(images=image, return_tensors="pt")
        inputs = {
            name: tensor.to(device, dtype=dtype) if tensor.is_floating_point() else tensor.to(device)
            for name, tensor in inputs.items()
        }

        options = {"max_new_tokens": args.max_new_tokens}
        if args.temperature is not None:
            if args.temperature == 0:
                options["do_sample"] = False
            else:
                options["do_sample"] = True
                options["temperature"] = args.temperature
        if args.seed is not None:
            torch.manual_seed(args.seed)

        log("[image-to-text] Generating text...")
        with torch.no_grad():
            output_ids = model.generate(
                **inputs,
                **options,
                streamer=StderrStreamer(tokenizer, skip_prompt=True, skip_special_tokens=True),
            )

        # Vision-language models return the chat prompt followed by the answer; captions keep their prompt as their start.
        new_ids = output_ids[0][inputs["input_ids"].shape[-1]:] if chat else output_ids[0]
        result = tokenizer.decode(new_ids, skip_special_tokens=True).strip()
    except Exception as e:
        traceback.print_exc()
        log(f"[image-to-text] Error: {type(e).__name__}: {e}")
        return EXIT_FAILED

    print(json.dumps({"text": result}), flush=True)
    log("[image-to-text] Done.")
    return EXIT_OK


def best_device(torch) -> str:
    if torch.cuda.is_available():
        return "cuda"
    if torch.backends.mps.is_available():
        return "mps"
    return "cpu"


if __name__ == "__main__":
    # torch starts multiprocessing helpers by re-running sys.executable, which in the frozen binary is this
    # script; freeze_support() lets those helper invocations run instead of reaching the argument parser.
    multiprocessing.freeze_support()
    sys.exit(main(sys.argv[1:]))

File: src/Runner/ImageToText.php
<?php

declare(strict_types=1);

namespace PhpLovesAi\Runner;

use PhpLovesAi\Binary\Tool;
use PhpLovesAi\Exception\BinaryNotInstalledException;
use PhpLovesAi\Exception\HomeDirectoryNotFoundException;
use PhpLovesAi\Exception\ImageNotFoundException;
use PhpLovesAi\Exception\ModelNotFoundException;
use PhpLovesAi\Exception\RunFailedException;
use PhpLovesAi\Exception\UnsupportedModelException;
use PhpLovesAi\Filesystem\Path;

/**
 * Describes images, or answers questions about them, with transformers models pulled into the project, through the
 * runner installed by `vendor/bin/loves-ai setup image-to-text`.
 *
 *     $caption = (new ImageToText())->generate('HuggingFaceTB/SmolVLM-256M-Instruct', __DIR__ . '/photo.jpg');
 *
 * Works with vision-language models (Hugging Face task "image-text-to-text", e.g. SmolVLM, Qwen2-VL, LLaVA), which
 * answer the prompt about the image, and captioning models ("image-to-text", e.g. BLIP, GIT, ViT-GPT2), for which a
 * prompt is the start of the caption.
 *
 * Input is passed to the model as-is: values it cannot handle make the run fail with a RunFailedException.
 */
final class ImageToText extends BinaryRunner
{
    /** A model the runner can load; used in error messages and hints. */
    public const EXAMPLE_MODEL = 'HuggingFaceTB/SmolVLM-256M-Instruct';

    /** Hugging Face models this runner can use. */
    public const COMPATIBLE_MODELS_URL = 'https://huggingface.co/models?pipeline_tag=image-text-to-text&library=transformers';

    public static function tool(): Tool
    {
        return Tool::ImageToText;
    }

    /**
     * @param string                        $model        Hugging Face model id, e.g. "HuggingFaceTB/SmolVLM-256M-Instruct"
     * @param string                        $imagePath    image file to read, e.g. "photo.jpg"; a leading "~" is expanded
     * @param string|null                   $prompt       question about the image for vision-language models (null asks
     *                                                    them to describe it), or the start of the caption for
     *                                                    captioning models
     * @param int|null                      $maxNewTokens maximum length of the text in tokens; null uses 256
     * @param float|null                    $temperature  randomness: 0 always picks the likeliest words; null uses the
     *                                                    model's default
     * @param int|null                      $seed         random seed, for reproducible text
     * @param string|null                   $device       torch device such as "cpu", "cuda" or "mps";
     *                                                    null picks the best available
     * @param (callable(string): void)|null $onOutput     receives the runner's progress output and the text as it is
     *                                                    generated
     *
     * @return string the description, caption or answer
     *
     * @throws ImageNotFoundException
     * @throws BinaryNotInstalledException
     * @throws ModelNotFoundException
     * @throws UnsupportedModelException when the model cannot read images or is not a transformers model
     * @throws HomeDirectoryNotFoundException
     * @throws RunFailedException
     */
    public function generate(
        string $model,
        string $imagePath,
        ?string $prompt = null,
        ?int $maxNewTokens = null,
        ?float $temperature = null,
        ?int $seed = null,
        ?string $device = null,
        ?callable $onOutput = null,
    ): string {
        $imagePath = Path::expandHome($imagePath);
        if (!is_file($imagePath) || !is_readable($imagePath)) {
            throw new ImageNotFoundException($imagePath);
        }

        $result = $this->run($model, [
            '--image' => $imagePath,
            '--prompt' => $prompt,
            '--max-new-tokens' => $maxNewTokens,
            '--temperature' => $temperature,
            '--seed' => $seed,
            '--device' => $device,
        ], $onOutput);

        if (!is_string($result['text'] ?? null)) {
            throw new RunFailedException(0, 'The runner did not report the generated text.');
        }

        return $result['text'];
    }

    /**
     * Besides what every transformers runner needs, the model must come with an image processor, which text-only
     * models lack.
     */
    protected function ensureModelIsSupported(string $model, string $modelPath): void
    {
        $problem = TransformersModel::problem($modelPath);

        if ($problem === null
            && !TransformersModel::has($modelPath, 'preprocessor_config.json')
            && !TransformersModel::has($modelPath, 'processor_config.json')
        ) {
            $problem = 'it cannot read images (it has no preprocessor_config.json). If it is a text model, use it with text-to-text.';
        }

        if ($problem === null) {
            return;
        }

        throw new UnsupportedModelException($model, $modelPath, sprintf(
            '%s cannot describe images: %s Use a transformers image-to-text model instead, e.g. %s (browse: %s).',
            $model,
            $problem,
            self::EXAMPLE_MODEL,
            self::COMPATIBLE_MODELS_URL,
        ));
    }
}
